### PR TITLE
feat(system): Gamut System Alpha

### PR DESCRIPTION
**Pull request by codecaaron**
## Overview

Partial Rebase of #953 

Checks in `gamut-system` for alpha testing within gamut.  Full information on the proposal is below.

### PR Checklist

- [x] Related to designs: All of them
- [x] Related to JIRA ticket: ABC-123
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

------------

# Full Description

## Why not Styled System?

[styled-system](https://styled-system.com/) does not support true type-safety creating a significant headache.

The closest level of type safety allowed by typescript[ is array level indexing](https://codesandbox.io/s/sparkling-glade-pl7jz?file=/src/App.tsx), at any level above that it is number or string only.  This is especially problematic for theme level scales, if we cannot detect that our theme shape has changed then we are prone to introducing bugs and encourage looser types.

## Features

- Completely customizable style props (with theme agnostic implementation);
- Flexible props that are both responsive and composable.
- Strict design system aware types with no manual fuss.
- Works with `emotion` and `styled-components` in preferred syntax.

## Getting Started

Gamut system is a styled-system inspired utility toolkit that focuses on making your design system
typesafe.

To use gamut-system, install gamut-system and your CSS in JS library of choice:

```sh
yarn add @codecademy/gamut-system @emotion/core @emotion/styled
```
## Initialize your system

The first thing you need to use `gamut-system` is to initialize your utility functions. This is necessary to generate
the correct types for your properties.

```tsx
import { system } from '@codecademy/gamut-system';

export const {
  typography,
  layout,
  spacing,
  border
  background
  color,
  flexbox,
  grid,
  properties,
  variants
} = system.create({});
```

## Create a component

Use your style props in your components:

```tsx
import styled from '@emotion/styled';
import { HandlerProps } from '@codecademy/gamut-system';
import { layout, spacing } from './system';

type BoxProps = HandlerProps<typeof layout> & HandlerProps<typeof spacing>;

export const Box = styled.div<BoxProps>`
  ${spacing}
  ${layout}
`;
```

## Style Functions

Blank diagram (1)

## Using your style functions 
Use your style props in your components:

```tsx
import styled from '@emotion/styled';
import { HandlerProps } from '@codecademy/gamut-system';
import { layout, spacing } from './system';

type BoxProps = HandlerProps<typeof layout> & HandlerProps<typeof spacing>;

export const Box = styled.div<BoxProps>`
  ${spacing}
  ${layout}
`;
```

Now you can use both `spacing` and `layout` props on your `<Box/>`.

```tsx
<Box padding="1rem" display="flex" />
```

### Responsive Props
By default, all `system` generated style functions accept responsive configurations.  Responsive configurations are two basic shapes a map of breakpoint aliases/values or an array of values:

```tsx
// Array Syntax
<Box width={[50, 100, 200]} />

// Object syntax
<Box margin={{ xs: 16, sm: 24, md: 32 }} />
```

## Composition

There may be instances where you would like to combine properties for ease of sharing or performance to batch their computations.  To make this easy we've added a compose function to combine returned style functions into a single one with the intersection of all of their props.

By default system will return a few groups of composed or group functions `typography`, `layout`, `spacing`,`border`, `background`, `color`, `flexbox`,`grid`.

Blank diagram (2)

However, we expose a `compose` function to combine any style function as necessary.

```tsx
import { compose } from '@codecademy/gamut-system';
import { fontSize, letterSpacing } from './system';

const myCoolComposed = compose(letterSpacing, fontSize);
````

The prop types of the resulting function will be:
```tsx
type MyCoolComposed = {
  letterSpacing?: <%TYPE> 
  fontSize?: <%TYPE>
};
```
## Variants

There may be cases where you want to change many props at the same time to achieve a different effect. Coordinating these different prop variations can be a bit frustrating to do manually, in both typescript and plain javascript. To make this easier we have `variant`, a function that is meant to make creating these combinations easy and scalable.

Blank diagram (3)

```tsx
import styled from '@emotion/styled';
import { system, HandlerProps } from '@codecademy/gamut-system';

export const { variants } = system.create({});

const buttonVariants = variants({
  variants: {
    primary: {
      borderStyle: 'solid',
      borderWidth: '1px',
      color: 'white',
      borderColor: 'darkblue',
      backgroundColor: 'blue',
    },
    secondary: {
      borderStyle: 'solid',
      borderWidth: '1px',
      color: 'white',
      borderColor: 'darkgreen',
      backgroundColor: 'green',
    },
  },
});

export type ButtonProps = HandlerProps<typeof ButtonVariants>;

export const Button = styled.button<ButtonProps>`
  ${buttonVariants}
`;
```

Using your variants:

```tsx
  <Button variant="primary" />

  <Button variant="secondary" />
```

There may be cases where you want to use a special key for your variants instead of the default `variant`. You may pass an arbitrary key for your prop to customize the prop type.

```tsx
const colorVariant = variants({
  key: 'colorVariant',
  variants: {
    primary: { color: 'blue' },
  },
});

const Text = styled.p`
  ${colorVariant}
`;

<Text colorVariant="primary" />;
```

This also allows you to use multiple variants in the same component to have combinatorial styles.

```tsx
import { HandlerProps } from '@codecademy/gamut-system';

import { boxVariants, colorVariants } from './myVariants';

type CardProps = HandlerProps<typeof boxVariants> &
  HandlerProps<typeof colorVariants>;

const Card = styled.div<CardProps>`
  ${boxVariants}
  ${colorVariants}
`;

<Card variant="primary" colorVariant="primary" />;
```

## Responsive Properties

Being able to configure your styles for multiple screen sizes is a common requirement. We make all style functions accept valid configurations of our responsive property spec.

Responsive configurations are two basic shape a map of breakpoint aliases / values or an array of values:

```tsx
export type MediaQueryArray<Value> = [
  Value?, // base
  Value?, // xs
  Value?, // sm
  Value?, // md
  Value?, // lg
  Value? // xl
];

export type MediaQueryMap<Value> = {
  base?: Value;
  xs?: Value;
  sm?: Value;
  md?: Value;
  lg?: Value;
  xl?: Value;
};

export type ResponsiveProp<Value> =
  | Value
  | MediaQueryArray<Value>
  | MediaQueryMap<Value>;
```

What this looks like in practice:

```tsx
const Box = styled.div`
  ${display}
`

// Array Syntax
<Box display={['none', ,'block']} />

// Alias Syntax
<Box display={{ base: 'none', sm: 'block' }} />
```

Output styles:

```scss
.#{emotionHash} {
  display: 'none';

  @media (min-width: 480px) {
    display: 'block';
  }
}
```

## Default Breakpoints

If you do not have a preference on the breakpoints that you use we default to this set:

| Alias | Mediaquery                   |
| ----- | ---------------------------- |
| xs    | `@media (min-width: 320px)`  |
| sm    | `@media (min-width: 480px)`  |
| md    | `@media (min-width: 768px)`  |
| lg    | `@media (min-width: 1024px)` |
| xl    | `@media (min-width: 1248px)` |

## Custom Breakpoints

If you would like to use your own, we allow you to specify 5 different breakpoints through the theme shape:

```tsx
import React from 'react';
import { system } from '@codecademy/gamut-system';
import { ThemeProvider } from 'emotion-theming';

const MyTheme = {
  // Other theme configurations
  breakpoints: {
    xs: '@media (max-width: 2px)';
    sm: '@media (max-width: 4px)';
    md: '@media (max-width: 8px)';
    lg: '@media (max-width: 16px)';
    xl: '@media (max-width: 32px)';
  }
}

export const {
  layout,
  spacing,
  typography,
  variant
} = system.withTheme<MyTheme>().create({});

export const Apps = ({ children }) => <ThemeProvider theme={MyTheme}>{children}</ThemeProvider>
```

## Customizing your System

By default calling system will return the base `CSSType` definitions for each rule and some extra transforms for quality of life. However,
there are many cases where we want more strict typings for our style functions or to have custom transformations. This can be configured
by creating a prop config: an immutable definition that defines the way the style functions should be constructed.

## Config Specifcation

- `propName`: Any defined prop in the system.
- `dependentProps`: An array of alternative props that you would like to group with the same style function.
- `type`: Declare the template type `standard` by default or `directional` for properties that have a shorthand for directional values.
- `transformValue`: There may be cases where you would like to transform or compute a property value into something more CSS
  friendly some examples may be adding units or looking up configurations. You may provide a `transformValue` function on your config to do this.
- `scale`: Either a typed array/object literal or a key of your theme that has those values.

```tsx
type AbstractPropertyConfig = {
  propName: PropAlias;
  dependentProps?: Readonly<string[]>;
  type?: 'standard' | 'directional';
  scale?: AbstractScales;
  transformValue?: TransformValue;
};
```

## Usage

To use this you must pass a subset of the base config and its property definitions (propGroups and handlers):

```tsx
export const { properties } = system.create({
  typography: {
    fontSize: {
      propName: 'fontSize',
      scale: {
        xs: '1rem',
        sm: '1.5rem',
        md: '2rem',
        lg: '3rem',
      },
    },
  },
});
```

The resulting types for your `fontSize` property and ALL consumers (`typography` and `variants`) will now be only the values specified as keys on your scale:

```tsx
// Compiles
<Text fontSize="xs" />

// Type Error
<Text fontSize="1rem" />
```

### Using with a theme

Using a theme can make this even more flexible.

```tsx

type ThemeShape = {
  fontSizes: { xs: string, sm: string, md: string, lg: string };
}


export const { properties }  = (system as ThemedSystem<ThemeShape>)({
  typography: {
    fontSize: {
      propName: 'fontSize',
      scale: 'fontSizes'
  },
});
```

In your react code:

```tsx
import React from 'react';
import { ThemeProvider } from 'emotion-theming';

const theme = {
  fontSizes: {
    xs: '1rem',
    sm: '1.5rem',
    md: '2rem',
    lg: '3rem',
  },
};

const App = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);
```

Your components will automatically look these values up from your theme. This also means the values of your keys can change depending on the context.
Making aliasing color combinations etc far easier.

## Initial Props
The preconfigured scales and API comes out of the box with some base properties to either be extended upon or added to.  We've added some composed style functions for grouped properties.

#### Spacing

| Prop          | CSS Property                | defaultTransform |
| ------------- | --------------------------- | ---------------- |
| margin        | margin                      |                  |
| marginX       | margin-left, margin-right   |                  |
| marginY       | margin-top, margin-bottom   |                  |
| marginTop     | margin-top                  |                  |
| marginBottom  | margin-bottom               |                  |
| marginLeft    | margin-left                 |                  |
| marginRight   | margin-right                |                  |
| padding       | padding                     |                  |
| paddingX      | padding-left, padding-right |                  |
| paddingY      | padding-top, padding-bottom |                  |
| paddingTop    | padding-top                 |                  |
| paddingBottom | padding-bottom              |                  |
| paddingLeft   | padding-left                |                  |
| paddingRight  | padding-right               |                  |

#### Typography

| Prop          | CSS Property   | defaultTransform |
| ------------- | -------------- | ---------------- |
| fontFamily    | font-family    |                  |
| fontWeight    | font-weight    |                  |
| fontSize      | font-size      |                  |
| lineHeight    | line-height    |                  |
| textAlign     | text-align     |                  |
| fontStyle     | font-style     |                  |
| letterSpacing | letter-spacing |                  |

#### Layout

| Prop         | CSS Property   | defaultTransform |
| ------------ | -------------- | ---------------- |
| display      | display        |                  |
| overflow     | overflow       |                  |
| overflowY    | overflow-y     |                  |
| overflowX    | overflow-x     |                  |
| width        | width          | `parseSize`      |
| minWidth     | min-width      | `parseSize`      |
| maxWidth     | max-width      | `parseSize`      |
| height       | height         | `parseSize`      |
| minHeight    | min-height     | `parseSize`      |
| maxHeight    | max-height     | `parseSize`      |
| verticalAign | vertical-align |                  |

#### Colors

| Prop            | CSS Property     | defaultTransform |
| --------------- | ---------------- | ---------------- |
| color           | color            |                  |
| borderColor     | border-color     |                  |
| backgroundColor | background-color |                  |

#### Background

| Prop               | CSS Property        | defaultTransform |
| ------------------ | ------------------- | ---------------- |
| backgroundImage    | background-image    |                  |
| backgroundSize     | background-size     |                  |
| backgroundRepeat   | background-repeat   |                  |
| backgroundPosition | background-position |                  |

#### Borders

| Prop              | CSS Property                          | defaultTransform |
| ----------------- | ------------------------------------- | ---------------- |
| borderWidth       | border-width                          |                  |
| borderWidthX      | border-left-width, border-right-width |                  |
| borderWidthY      | border-top-width, border-bottom-width |                  |
| borderWidthTop    | border-top-width                      |                  |
| borderWidthBottom | border-bottom-width                   |                  |
| borderWidthLeft   | border-left-width                     |                  |
| borderWidthRight  | border-right-width                    |                  |
| borderRadius      | border-radius                         |                  |
| borderStyle       | border-style                          |                  |

#### Position

| Prop     | CSS Property | defaultTransform |
| -------- | ------------ | ---------------- |
| position | position     |                  |
| top      | top          | `parseSize`      |
| right    | right        | `parseSize`      |
| bottom   | bottom       | `parseSize`      |
| left     | left         | `parseSize`      |
| zIndex   | zIndex       |                  |

#### Shadow

| Prop       | CSS Property | defaultTransform |
| ---------- | ------------ | ---------------- |
| boxShadow  | box-shadow   |                  |
| textShadow | text-shadow  |                  |

#### Flex

| Prop           | CSS Property    | defaultTransform |
| -------------- | --------------- | ---------------- |
| flexBasis      | flex-basis      | `parseSize`      |
| flexDirection  | flex-direction  |                  |
| flexWrap       | flex-wrap       |                  |
| flexShrink     | flex-shrink     |                  |
| flexGrow       | flex-grow       |                  |
| order          | order           |                  |
| justifySelf    | justify-self    |                  |
| justifyContent | justify-content |                  |
| justifyItems   | justify-items   |                  |
| alignSelf      | align-self      |                  |
| alignItems     | align-items     |                  |
| alignContent   | align-content   |                  |

#### Grid

| Prop                | CSS Property          | defaultTransform |
| ------------------- | --------------------- | ---------------- |
| gridAutoColumns     | grid-auto-columns     |                  |
| gridAutoRows        | grid-auto-rows        |                  |
| gridTemplateColumns | grid-template-columns |                  |
| gridTemplateRows    | grid-template-rows    |                  |
| gridAutoFlow        | grid-auto-flow        |                  |
| rowGap              | grid-row-gap          |                  |
| columnGap           | grid-column-gap       |                  |
| gridColumn          | grid-column           |                  |
| gridRow             | grid-row              |                  |
| gridColumnStart     | grid-column-start     |                  |
| gridRowStart        | grid-row-start        |                  |
| gridColumnEnd       | grid-column-end       |                  |
| gridRowEnd          | grid-row-end          |                  |

### CODE REVIEW

File: packages/gamut-system/README.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Gamut System
```
Comment (JoshuaKGoldberg) [inline packages/gamut-system/README.md:1]: Were you going to rename this to Variance? 😁

**Review by JoshuaKGoldberg — APPROVED** [review]
😍

**Review by jakemhiller — APPROVED** [review]

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-system@0.1.0-alpha.15dfd2.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5fa95e182629f4d7304d5570--gamut-preview.netlify.app>https://5fa95e182629f4d7304d5570--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/5fa95e182629f4d7304d5570>Deploy Logs</a></p>

**Pull request merged**